# Объекты даты и времени
 - datetime
 - time
 - calendar
 
 2 типа объектов:
 - naive
 - aware
 
 Важно правильно работать со временем, особенно в веб-приложениях

For applications requiring aware objects, datetime and time objects have an optional time zone information attribute, tzinfo, that can be set to an instance of a subclass of the abstract tzinfo class. 

In [14]:
import time

In [15]:
dir(time)

['CLOCK_BOOTTIME',
 'CLOCK_MONOTONIC',
 'CLOCK_MONOTONIC_RAW',
 'CLOCK_PROCESS_CPUTIME_ID',
 'CLOCK_REALTIME',
 'CLOCK_TAI',
 'CLOCK_THREAD_CPUTIME_ID',
 '_STRUCT_TM_ITEMS',
 '__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'altzone',
 'asctime',
 'clock_getres',
 'clock_gettime',
 'clock_gettime_ns',
 'clock_settime',
 'clock_settime_ns',
 'ctime',
 'daylight',
 'get_clock_info',
 'gmtime',
 'localtime',
 'mktime',
 'monotonic',
 'monotonic_ns',
 'perf_counter',
 'perf_counter_ns',
 'process_time',
 'process_time_ns',
 'pthread_getcpuclockid',
 'sleep',
 'strftime',
 'strptime',
 'struct_time',
 'thread_time',
 'thread_time_ns',
 'time',
 'time_ns',
 'timezone',
 'tzname',
 'tzset']

In [17]:
time.time()

1655137525.8814912

In [18]:
help(time.time)

Help on built-in function time in module time:

time(...)
    time() -> floating point number
    
    Return the current time in seconds since the Epoch.
    Fractions of a second may be present if the system clock provides them.



In [16]:
time.CLOCK_BOOTTIME

7

In [21]:
start = time.time()
for i in range(1_000_000):
    i**10
print(time.time() - start)

0.3229789733886719


In [24]:
time.localtime(0)

time.struct_time(tm_year=1970, tm_mon=1, tm_mday=1, tm_hour=3, tm_min=0, tm_sec=0, tm_wday=3, tm_yday=1, tm_isdst=0)

In [3]:
import datetime

In [ ]:
# from datetime import datetime

Диапазоны лет..

In [25]:
datetime.MINYEAR

1

In [26]:
datetime.MAXYEAR

9999

In [27]:
cdt = datetime.datetime.now()

In [28]:
print(cdt) # __str__() # __repr__()

2022-06-13 19:31:34.620376


In [30]:
cdt

datetime.datetime(2022, 6, 13, 19, 31, 34, 620376)

In [33]:
datetime.datetime(2022, 5, 10, 19, 31, 34, 620376)

datetime.datetime(2022, 5, 10, 19, 31, 34, 620376)

In [ ]:
type(cdt)

In [34]:
date = datetime.datetime.now().date() # текущая дата и время, -> дата

In [35]:
type(date)

datetime.date

In [40]:
dir(date)

['__add__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__ne__',
 '__new__',
 '__radd__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rsub__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 'ctime',
 'day',
 'fromisocalendar',
 'fromisoformat',
 'fromordinal',
 'fromtimestamp',
 'isocalendar',
 'isoformat',
 'isoweekday',
 'max',
 'min',
 'month',
 'replace',
 'resolution',
 'strftime',
 'timetuple',
 'today',
 'toordinal',
 'weekday',
 'year']

In [39]:
date.year

2022

In [ ]:
date

In [ ]:
type(date)

In [44]:
d = datetime.datetime(2021, 1, 1)

In [42]:
d

datetime.datetime(2021, 1, 4, 0, 0)

In [ ]:
dir(datetime)

In [ ]:
datetime.date(2019, 12, 30)

```python
    #new in python3.9
```

[zoneinfo module in standard lib](https://docs.python.org/3/library/zoneinfo.html#module-zoneinfo)

In [1]:
import pytz # необходимо установить (pip install pytz) (old/original way)

In [4]:
from zoneinfo import ZoneInfo

In [5]:
dt = datetime.datetime(2021, 4, 23, 12, tzinfo=ZoneInfo("Europe/Minsk"))

In [6]:
dt

datetime.datetime(2021, 4, 23, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='Europe/Minsk'))

In [7]:
dt.utcoffset()

datetime.timedelta(seconds=10800)

In [8]:
10800/60/60

3.0

In [10]:
dt

datetime.datetime(2021, 4, 23, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='Europe/Minsk'))

In [9]:
timezone = pytz.timezone("America/Belem")

In [10]:
tz_kashgar = pytz.timezone('Asia/Kashgar')

In [11]:
pytz.all_timezones

['Africa/Abidjan',
 'Africa/Accra',
 'Africa/Addis_Ababa',
 'Africa/Algiers',
 'Africa/Asmara',
 'Africa/Asmera',
 'Africa/Bamako',
 'Africa/Bangui',
 'Africa/Banjul',
 'Africa/Bissau',
 'Africa/Blantyre',
 'Africa/Brazzaville',
 'Africa/Bujumbura',
 'Africa/Cairo',
 'Africa/Casablanca',
 'Africa/Ceuta',
 'Africa/Conakry',
 'Africa/Dakar',
 'Africa/Dar_es_Salaam',
 'Africa/Djibouti',
 'Africa/Douala',
 'Africa/El_Aaiun',
 'Africa/Freetown',
 'Africa/Gaborone',
 'Africa/Harare',
 'Africa/Johannesburg',
 'Africa/Juba',
 'Africa/Kampala',
 'Africa/Khartoum',
 'Africa/Kigali',
 'Africa/Kinshasa',
 'Africa/Lagos',
 'Africa/Libreville',
 'Africa/Lome',
 'Africa/Luanda',
 'Africa/Lubumbashi',
 'Africa/Lusaka',
 'Africa/Malabo',
 'Africa/Maputo',
 'Africa/Maseru',
 'Africa/Mbabane',
 'Africa/Mogadishu',
 'Africa/Monrovia',
 'Africa/Nairobi',
 'Africa/Ndjamena',
 'Africa/Niamey',
 'Africa/Nouakchott',
 'Africa/Ouagadougou',
 'Africa/Porto-Novo',
 'Africa/Sao_Tome',
 'Africa/Timbuktu',
 'Africa/

In [12]:
tz_minsk = pytz.timezone("Europe/Minsk")

In [13]:
tz_minsk

<DstTzInfo 'Europe/Minsk' LMT+1:50:00 STD>

In [14]:
pytz.VERSION

'2022.1'

In [15]:
timezone


<DstTzInfo 'America/Belem' LMT-1 day, 20:46:00 STD>

In [ ]:
d = datetime.datetime(2019, 3, 23, 13, 10, 52,tzinfo=timezone)

In [ ]:
d

In [16]:
ct_minsk = datetime.datetime(2020, 5, 15, 12, 0, 0,tzinfo=tz_minsk)

In [17]:
ct_minsk

datetime.datetime(2020, 5, 15, 12, 0, tzinfo=<DstTzInfo 'Europe/Minsk' LMT+1:50:00 STD>)

In [18]:
ct_minsk.utcoffset()

datetime.timedelta(seconds=6600)

In [ ]:
ct_minsk

In [19]:
6600/60 # 180

110.0

# Из документации pytz
## правильное добавление информации о зоне

In [20]:
ct_minsk.utcoffset() # для Минска сдвиг 6600 сек - ошибка

datetime.timedelta(seconds=6600)

In [21]:
ct_minsk # нет инф о зоне - ваша задача создать правильный объект

datetime.datetime(2020, 5, 15, 12, 0, tzinfo=<DstTzInfo 'Europe/Minsk' LMT+1:50:00 STD>)

In [22]:
ct_minsk = datetime.datetime.now()

In [23]:
ct_minsk

datetime.datetime(2022, 6, 13, 19, 53, 54, 585107)

In [24]:
tz_minsk

<DstTzInfo 'Europe/Minsk' LMT+1:50:00 STD>

In [25]:
d_minsk = tz_minsk.localize(ct_minsk) # так рекомендует документация pytz

In [26]:
d_minsk

datetime.datetime(2022, 6, 13, 19, 53, 54, 585107, tzinfo=<DstTzInfo 'Europe/Minsk' +03+3:00:00 STD>)

In [27]:
d_minsk.utcoffset() # теперь оффсет правильный

datetime.timedelta(seconds=10800)

In [28]:
10800/60


180.0

In [37]:

utc_time = datetime.datetime.utcnow()

In [38]:
utc_time # опять наивное время 

datetime.datetime(2022, 6, 13, 16, 59, 56, 294420)

In [35]:
utc_time.utcoffset()

datetime.timedelta(0)

In [40]:
utc_tz = pytz.timezone("UTC")

In [41]:
utc_time = utc_tz.localize(utc_time)

In [42]:
utc_time

datetime.datetime(2022, 6, 13, 16, 59, 56, 294420, tzinfo=<UTC>)

In [43]:
utc_time.astimezone(tz_kashgar)

datetime.datetime(2022, 6, 13, 22, 59, 56, 294420, tzinfo=<DstTzInfo 'Asia/Kashgar' +06+6:00:00 STD>)

In [44]:
utc_time.astimezone(tz_minsk)

datetime.datetime(2022, 6, 13, 19, 59, 56, 294420, tzinfo=<DstTzInfo 'Europe/Minsk' +03+3:00:00 STD>)

In [45]:
utc_time.astimezone(tz_kashgar)- utc_time.astimezone(tz_minsk) # = 0 как доктор прописал

datetime.timedelta(0)

## Date utils
```bash
pip install python-dateutil
```

In [ ]:
from dateutil import tz
now = datetime.datetime.now(tz=tz.tzlocal())

In [ ]:
now.tzname()

In [ ]:
tz_utc = tz.UTC
now_utc = datetime.datetime.now(tz=tz_utc)

In [ ]:
now_utc.tzname()

In [ ]:
now_utc.utcoffset()

In [ ]:
now.utcoffset()

In [ ]:
now

In [ ]:
from dateutil.zoneinfo import get_zonefile_instance
zonenames = list(get_zonefile_instance().zones)

In [ ]:
zonenames

In [ ]:
tz_kz = tz.gettz("Asia/Almaty")

In [ ]:
now_kz = datetime.datetime.now(tz=tz_kz)

In [ ]:
now_kz

In [ ]:
now_kz.utcoffset()

## Date
```python
datetime.date # class
```

In [ ]:
date = datetime.date(2017, 1, 15)

In [ ]:
type(date)

In [ ]:
date

## Time
```python
datetime.time # class
```

In [46]:
time = datetime.time(23, 59, 59)

In [47]:
print(time) # no tzinfo

23:59:59


## Timedelta

In [48]:
td = datetime.timedelta(seconds=3600)

In [50]:
td

datetime.timedelta(seconds=3600)

In [51]:
cdt = datetime.datetime.now()

In [52]:
cdt

datetime.datetime(2022, 6, 13, 20, 16, 41, 456868)

In [53]:
cdt + td

datetime.datetime(2022, 6, 13, 21, 16, 41, 456868)

In [54]:
str(td)

'1:00:00'

timedelta(days=0, seconds=0, microseconds=0, milliseconds=0, minutes=0, hours=0, weeks=0)

In [55]:
dt = datetime.timedelta(weeks=3, hours=4, days=3)

In [56]:
dir(dt)

['__abs__',
 '__add__',
 '__bool__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__pos__',
 '__radd__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rmod__',
 '__rmul__',
 '__rsub__',
 '__rtruediv__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 'days',
 'max',
 'microseconds',
 'min',
 'resolution',
 'seconds',
 'total_seconds']

In [57]:
dt_minutes = datetime.timedelta(minutes=3)

In [58]:
dt_minutes.days

0

In [ ]:
td = datetime.timedelta(10000)

In [59]:
ussr = datetime.date(1991,8,21)

In [60]:
today = datetime.datetime.now().date()

In [61]:
print(ussr - today)

-11254 days, 0:00:00


In [65]:
tn1 = datetime.datetime.now()

In [66]:
tn2 = datetime.datetime.now()

In [67]:
tn1 - tn2

datetime.timedelta(days=-1, seconds=86399, microseconds=484674)

In [68]:
dir(datetime.timedelta)

['__abs__',
 '__add__',
 '__bool__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__pos__',
 '__radd__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rmod__',
 '__rmul__',
 '__rsub__',
 '__rtruediv__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 'days',
 'max',
 'microseconds',
 'min',
 'resolution',
 'seconds',
 'total_seconds']

In [69]:
rb + 3

NameError: name 'rb' is not defined

In [ ]:
print( f'Дней: {diff.total_seconds()}')

```bash
# аналог
from dateutil.relativedelta import relativedelta
```

### Преобразования из/в строку

In [70]:
tn = datetime.datetime.now()

In [71]:
tn

datetime.datetime(2022, 6, 13, 20, 20, 52, 752693)

In [72]:
str(tn)

'2022-06-13 20:20:52.752693'

In [ ]:
type(tn)

In [73]:
tn

datetime.datetime(2022, 6, 13, 20, 20, 52, 752693)

In [76]:
str_tn = tn.strftime("%d.%m.%Y %a %B %H:just some text %M:%S") # time to string

In [77]:
str_tn

'13 Mon June 20:just some text 20:52'

In [79]:
tn.strftime("%d.%m.%Y")

'13.06.2022'

In [85]:
time_as_string = '18.03.1819 123456 13:27'

In [86]:
time_obj = datetime.datetime.strptime(time_as_string, '%d.%m.%Y 123456 %H:%M')

In [87]:
time_obj

datetime.datetime(1819, 3, 18, 13, 27)

In [88]:
time_obj = datetime.datetime.strptime

In [89]:
time_obj = datetime.datetime.strptime

In [90]:
time_obj1 = datetime.datetime(day=1, month=1, year=2001)

In [91]:
time_as_string

'18.03.1819 123456 13:27'

In [92]:
time_obj = datetime.datetime.strptime(time_as_string, "%d.%m.%y %H:%M")

ValueError: time data '18.03.1819 123456 13:27' does not match format '%d.%m.%y %H:%M'

In [93]:
time_obj

<function datetime.strptime>

In [94]:
time_obj

<function datetime.strptime>

In [ ]:
tn

In [ ]:
ct_minsk

In [ ]:
now_kz.strftime('%a %w %b %Z %z')


| Directive | Meaning |	Example	| Notes |
| :--- | :--- | :--- | :--- |
| %a |	Weekday as locale’s abbreviated name. | Sun, Mon, …, Sat (en_US); So, Mo, …, Sa (de_DE) | (1) |
 |%A |	Weekday as locale’s full name. | Sunday, Monday, …, Saturday (en_US); Sonntag, Montag, …, Samstag (de_DE) | (1) |
|%w	|Weekday as a decimal number, where 0 is Sunday and 6 is Saturday.	|0, 1, …, 6| |
|%d	|Day of the month as a zero-padded decimal number.	|01, 02, …, 31	| | 
|%b	|Month as locale’s abbreviated name.	|Jan, Feb, …, Dec (en_US); Jan, Feb, …, Dez (de_DE) | (1)|
|%B	|Month as locale’s full name.	| January, February, …, December (en_US); Januar, Februar, …, Dezember (de_DE) | (1)|
|%m	|Month as a zero-padded decimal number.|	01, 02, …, 12| |	 
|%y	|Year without century as a zero-padded decimal number.	|00, 01, …, 99	| | 
|%Y	|Year with century as a decimal number.	|0001, 0002, …, 2013, 2014, …, 9998, 9999|	(2)|
|%H	|Hour (24-hour clock) as a zero-padded decimal number.	|00, 01, …, 23	 | |
|%I	|Hour (12-hour clock) as a zero-padded decimal number.	|01, 02, …, 12	 | |
|%p	|Locale’s equivalent of either AM or PM. |AM, PM (en_US); am, pm (de_DE)| (1), (3)
|%M	|Minute as a zero-padded decimal number.	|00, 01, …, 59	| | 
|%S	|Second as a zero-padded decimal number.	|00, 01, …, 59	|(4)|
|%f	|Microsecond as a decimal number, zero-padded on the left.	|000000, 000001, …, 999999| (5)|
|%z	|UTC offset in the form ±HHMM[SS[.ffffff]] (empty string if the object is naive).|(empty), +0000, -0400, +1030, +063415, -030712.345216	|(6)|
|%Z	|Time zone name (empty string if the object is naive).|	(empty), UTC, EST, CST| |	 
|%j	|Day of the year as a zero-padded decimal number.	|001, 002, …, 366| |	 
|%U	|Week number of the year (Sunday as the first day of the week) as a zero padded decimal number. All days in a new year preceding the first Sunday are considered to be in week 0.	|0, 01, …, 53|	(7)|
|%W	|Week number of the year (Monday as the first day of the week) as a decimal number. All days in a new year preceding the first Monday are considered to be in week 0.|	00, 01, …, 53|(7)|
|%c	|Locale’s appropriate date and time representation.	|Tue Aug 16 21:30:00 1988 (en_US); Di 16 Aug 21:30:00 1988 (de_DE)|(1)|
|%x	|Locale’s appropriate date representation.	|08/16/88 (None); 08/16/1988 (en_US); 16.08.1988 (de_DE)|(1)|
|%X	|Locale’s appropriate time representation.	|21:30:00 (en_US); 21:30:00 (de_DE)|(1)|
|%%	|A literal '%' character.	|%	 | |

## Calendar

In [95]:
import calendar

In [96]:
wd = calendar.Calendar()

In [97]:
for d in wd.iterweekdays():
    print(d)

0
1
2
3
4
5
6


In [98]:
for d in wd.itermonthdates(month=2, year=2014):
    print(d)

2014-01-27
2014-01-28
2014-01-29
2014-01-30
2014-01-31
2014-02-01
2014-02-02
2014-02-03
2014-02-04
2014-02-05
2014-02-06
2014-02-07
2014-02-08
2014-02-09
2014-02-10
2014-02-11
2014-02-12
2014-02-13
2014-02-14
2014-02-15
2014-02-16
2014-02-17
2014-02-18
2014-02-19
2014-02-20
2014-02-21
2014-02-22
2014-02-23
2014-02-24
2014-02-25
2014-02-26
2014-02-27
2014-02-28
2014-03-01
2014-03-02


In [99]:
wd.monthdatescalendar(1978, 10)

[[datetime.date(1978, 9, 25),
  datetime.date(1978, 9, 26),
  datetime.date(1978, 9, 27),
  datetime.date(1978, 9, 28),
  datetime.date(1978, 9, 29),
  datetime.date(1978, 9, 30),
  datetime.date(1978, 10, 1)],
 [datetime.date(1978, 10, 2),
  datetime.date(1978, 10, 3),
  datetime.date(1978, 10, 4),
  datetime.date(1978, 10, 5),
  datetime.date(1978, 10, 6),
  datetime.date(1978, 10, 7),
  datetime.date(1978, 10, 8)],
 [datetime.date(1978, 10, 9),
  datetime.date(1978, 10, 10),
  datetime.date(1978, 10, 11),
  datetime.date(1978, 10, 12),
  datetime.date(1978, 10, 13),
  datetime.date(1978, 10, 14),
  datetime.date(1978, 10, 15)],
 [datetime.date(1978, 10, 16),
  datetime.date(1978, 10, 17),
  datetime.date(1978, 10, 18),
  datetime.date(1978, 10, 19),
  datetime.date(1978, 10, 20),
  datetime.date(1978, 10, 21),
  datetime.date(1978, 10, 22)],
 [datetime.date(1978, 10, 23),
  datetime.date(1978, 10, 24),
  datetime.date(1978, 10, 25),
  datetime.date(1978, 10, 26),
  datetime.date(197

In [100]:




wd.monthdayscalendar(1978, 10)

[[0, 0, 0, 0, 0, 0, 1],
 [2, 3, 4, 5, 6, 7, 8],
 [9, 10, 11, 12, 13, 14, 15],
 [16, 17, 18, 19, 20, 21, 22],
 [23, 24, 25, 26, 27, 28, 29],
 [30, 31, 0, 0, 0, 0, 0]]

In [ ]:
dir(wd)

In [ ]:
dir(wd)

In [ ]:
from dateutil import tz

In [ ]:
tz.gettz("Europe/Minsk")

In [ ]:
now = datetime.datetime.now()

In [ ]:
now

In [ ]:
tz_minsk = tz.gettz("Europe/Minsk")

In [ ]:
now.replace(tzinfo=tz_minsk)

In [ ]:
now = now.replace(tzinfo=tz_minsk)

In [ ]:
now

In [ ]:
now.utcoffset()